File: Marston and Morgan's Territory Puzzle/sol.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef vector< vector<ll> > graph;

ll n, m;
graph undirectedGraph;
vector< ll> isVisited, even, odd;
bool isValid = true;

void dfs(ll src, ll parent) {
	if (isVisited[parent] & 1) {
		isVisited[src] = 2;
		odd.push_back(src);
	}
	else {
		isVisited[src] = 1;
		even.push_back(src);
	}
	for (auto it: undirectedGraph[src]) {
		if (!isVisited[it]) {
			dfs(it, src);
		} else if ((isVisited[it] & 1) == isVisited[src] & 1) {
			isValid = false;
		}
	}
}

void solve() {
	cin >> n >> m;
	undirectedGraph.resize(n + 1);
	for (int i =0; i< m; ++i) {
		ll u, v;
		cin >> u >> v;
		undirectedGraph[u].push_back(v);
		undirectedGraph[v].push_back(u);
	}
	isVisited.resize(n+1);
	for (int i=1; i<=n; ++i) {
		if ((!isVisited[i]) && (undirectedGraph[i].size()))
			dfs(i, i);
	}
	if(!isValid) {
		return void(cout << -1);
	}
	cout << odd.size()<<'\n';
	for (auto it: odd) cout<<it<<' ';
	cout<<'\n';
	cout<<even.size()<<'\n';
	for (auto it:even) cout<<it<<' ';
}

signed main() {
	int tt = 1;
	while (tt--) {
		solve();
		cout<<endl;
	}
	return 0;
}

File: Marston and Morgan's Territory Puzzle/sol.py
from collections import defaultdict

def dfs(src, par):
    global isVisited, graph, isValid, even, odd
    
    if isVisited[par] & 1:
        isVisited[src] = 2
        odd.append(src)
    else:
        isVisited[src] = 1
        even.append(src)

    for it in graph[src]:
        if not isVisited[it]:
            dfs(it, src)
        elif isVisited[it] & 1 == isVisited[src] & 1:
            isValid = False

def solve():
    global n, m, graph, isVisited, isValid, even, odd
    
    n, m = map(int, input().split())
    graph = defaultdict(list)
    for _ in range(m):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)

    isVisited = [0] * (n + 1)
    for i in range(1, n + 1):
        if not isVisited[i] and graph[i]:
            dfs(i, i)

    if not isValid:
        print(-1)
        return

    print(len(odd))
    print(*odd)
    print(len(even))
    print(*even)

if __name__ == "__main__":
    tt = 1
    while tt > 0:
        n, m = 0, 0
        graph = defaultdict(list)
        isVisited, even, odd = [], [], []
        isValid = True
        solve()
        print()
        tt -= 1
